File: src/main/java/io/github/dqualizer/dqapi/exceptions/GlobalExceptionHandler.java
package io.github.dqualizer.dqapi.exceptions;


import org.springframework.dao.DuplicateKeyException;
import org.springframework.http.HttpStatus;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ProblemDetail;
import org.springframework.http.ResponseEntity;
import org.springframework.web.ErrorResponse;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
public class GlobalExceptionHandler {
    @ExceptionHandler(NotFoundException.class)
    public ResponseEntity<?> handleNotFoundException(NotFoundException ex) {
        return new ResponseEntity<>(ex.getMessage(), HttpStatus.NOT_FOUND);
    }

    @ExceptionHandler(DuplicateEntityException.class)
    public ResponseEntity<?> handleDuplicateKeyException(DuplicateEntityException ex) {
        ErrorResponse error = ErrorResponse.create(ex, HttpStatus.CONFLICT, ex.getMessage());
        return new ResponseEntity<>(error, HttpStatus.CONFLICT);
    }
}

File: src/main/kotlin/io/github/dqualizer/dqapi/exceptions/GlobalExceptionHandler.kt
package io.github.dqualizer.dqapi.exceptions

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.ErrorResponse
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler


@ControllerAdvice
class GlobalExceptionHandler {
  @ExceptionHandler(NotFoundException::class)
  fun handleNotFoundException(ex: NotFoundException): ResponseEntity<*> {
    return ResponseEntity(ex.message, HttpStatus.NOT_FOUND)
  }

  @ExceptionHandler(DuplicateEntityException::class)
  fun handleDuplicateKeyException(ex: DuplicateEntityException): ResponseEntity<*> {
    val error = ErrorResponse.create(
      ex, HttpStatus.CONFLICT,
      ex.message!!
    )
    return ResponseEntity(error, HttpStatus.CONFLICT)
  }
}
